src/core/exception.d

```diff
@@ -416,12 +427,65 @@ deprecated void setAssertHandler( AssertHandler h ) @trusted nothrow @nogc
     assertHandler = h;
 }
 
+/*
+Returns a perThreadInstance object of class type `C`, i.e. repeated calls to
+`perThreadInstance!C` within a given thread return the same instance. For that
+reason, no constructor parameters are accepted; `C` must allow default
+construction.
+
+Different threads own different instances of `C`.
+
+The destructor of the perThreadInstance object is called during application's
+exit.
+
+Returns: A thread-local instance of type `C`.
+
+Throws: If the default constructor of `C` throws.
+*/
+private C perThreadInstance(C)() if (is(C == class) && is(typeof(new C)))
+{
+    static size_t[1 + (__traits(classInstanceSize, C) - 1) / size_t.sizeof] b;
+    static bool initialized = false;
+    auto result = () @trusted { return cast(C) b.ptr; }();
+    if (!initialized)
+    {
+        assert(typeid(C).initializer.length == b.length * size_t.sizeof);
+        import core.stdc.string;
+        () @trusted
+        {
+            memcpy(b.ptr, typeid(C).initializer.ptr, b.length);
```

Comment:
In theory the following could be more efficient `auto ci = typeid(C); b[0 .. ci.initializer.length] = ci.initializer[]` (assuming `b` is of type `ubyte[]`), because the compiler knows at compile time `typeid(C).initializer.length` and can make a better decision about the codegen (e.g. potentially inline the memcpy for smaller types).